doc/source/train/examples/pytorch/distributing-pytorch/README.ipynb

```diff
@@ -0,0 +1,599 @@
+    "\n",
+    "    # Get dataloaders inside the worker training function\n",
+    "    train_dataloader, test_dataloader = get_dataloaders(batch_size=batch_size)\n",
+    "\n",
+    "    # [1] Prepare Dataloader for distributed training\n",
+    "    # The prepare_data_loader method shards the dataset among workers and\n",
+    "    # moves batches to the correct device\n",
+    "    # ===================================================================\n",
+    "    train_dataloader = ray.train.torch.prepare_data_loader(train_dataloader)\n",
+    "    test_dataloader = ray.train.torch.prepare_data_loader(test_dataloader)\n",
+    "\n",
+    "    model = VisionTransformer(\n",
+    "        image_size=32,   # CIFAR-10 image size is 32x32\n",
+    "        patch_size=4,    # Patch size is 4x4\n",
+    "        num_layers=12,   # Number of transformer layers\n",
+    "        num_heads=8,     # Number of attention heads\n",
+    "        hidden_dim=384,  # Hidden size (can be adjusted)\n",
+    "        mlp_dim=768,     # MLP dimension (can be adjusted)\n",
+    "        num_classes=10   # CIFAR-10 has 10 classes\n",
+    "    )\n",
+    "\n",
+    "    # [2] Prepare and wrap your model with DistributedDataParallel\n",
+    "    # The prepare_model method moves the model to the correct GPU/CPU device\n",
+    "    # ======================================================================\n",
+    "    model = ray.train.torch.prepare_model(model)\n",
+    "\n",
+    "    loss_fn = nn.CrossEntropyLoss()\n",
+    "    optimizer = torch.optim.AdamW(model.parameters(), lr=lr, weight_decay=1e-2)\n",
+    "\n",
+    "    # Model training loop\n",
+    "    for epoch in range(epochs):\n",
+    "        if ray.train.get_context().get_world_size() > 1:\n",
+    "            # Required for the distributed sampler to shuffle properly across epochs.\n",
+    "            train_dataloader.sampler.set_epoch(epoch)\n",
+    "\n",
+    "        model.train()\n",
+    "        for X, y in tqdm(train_dataloader, desc=f\"Train Epoch {epoch}\"):\n",
+    "            pred = model(X)\n",
+    "            loss = loss_fn(pred, y)\n",
+    "\n",
+    "            optimizer.zero_grad()\n",
+    "            loss.backward()\n",
+    "            optimizer.step()\n",
+    "\n",
+    "        model.eval()\n",
+    "        test_loss, num_correct, num_total = 0, 0, 0\n",
+    "        with torch.no_grad():\n",
+    "            for X, y in tqdm(test_dataloader, desc=f\"Test Epoch {epoch}\"):\n",
+    "                pred = model(X)\n",
+    "                loss = loss_fn(pred, y)\n",
+    "\n",
+    "                test_loss += loss.item()\n",
+    "                num_total += y.shape[0]\n",
+    "                num_correct += (pred.argmax(1) == y).sum().item()\n",
+    "\n",
+    "        test_loss /= len(train_dataloader)\n",
+    "        accuracy = num_correct / num_total\n",
+    "\n",
+    "        # [3] Report metrics to Ray Train\n",
```

Comment:
(Optional) Report checkpoints and attached metrics to Ray Train

Reply:
done.